beacon_node/store/src/hdiff.rs

```diff
@@ -0,0 +1,489 @@
+    /// - Layer 0: 3000003 - (3000003 mod 2^21) = 2097152
+    /// - Layer 1: 3000003 - (3000003 mod 2^13) = 2998272
+    /// - Layer 2: 3000003 - (3000003 mod 2^5)  = 3000000
+    ///
+    /// Layer 0 is full state snaphost, apply layer 1 diff, then apply layer 2 diff and then replay
+    /// blocks 3,000,001 to 3,000,003.
+    pub exponents: Vec<u8>,
+}
+
+impl FromStr for HierarchyConfig {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, String> {
+        let exponents = s
+            .split(',')
+            .map(|s| {
+                s.parse()
+                    .map_err(|e| format!("invalid hierarchy-exponents: {e:?}"))
+            })
+            .collect::<Result<Vec<u8>, _>>()?;
+
+        if exponents.windows(2).any(|w| w[0] >= w[1]) {
+            return Err("hierarchy-exponents must be in ascending order".to_string());
+        }
+
+        Ok(HierarchyConfig { exponents })
+    }
+}
+
+#[derive(Debug)]
+pub struct HierarchyModuli {
+    moduli: Vec<u64>,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum StorageStrategy {
+    ReplayFrom(Slot),
+    DiffFrom(Slot),
+    Snapshot,
+}
+
+/// Hierarchical diff output and working buffer.
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct HDiffBuffer {
+    state: Vec<u8>,
+    balances: Vec<u64>,
+}
+
+/// Hierarchical state diff.
+///
+/// Splits the diff into two data sections:
+///
+/// - **balances**: The balance of each active validator is almost certain to change every epoch.
+///   So this is the field in the state with most entropy. However the balance changes are small.
+///   We can optimize the diff significantly by computing the balance difference first and then
+///   compressing the result to squash those leading zero bytes.
+///
+/// - **everything else**: Instead of trying to apply heuristics and be clever on each field,
+///   running a generic binary diff algorithm on the rest of fields yields very good results. With
+///   this strategy the HDiff code is easily mantainable across forks, as new fields are covered
+///   automatically. xdelta3 algorithm showed diff compute and apply times of ~200 ms on a mainnet
+///   state from Apr 2023 (570k indexes), and a 92kB diff size.
+#[derive(Debug, Encode, Decode)]
+pub struct HDiff {
+    state_diff: BytesDiff,
+    balances_diff: CompressedU64Diff,
+}
+
+#[derive(Debug, Encode, Decode)]
+pub struct BytesDiff {
+    bytes: Vec<u8>,
+}
+
+#[derive(Debug, Encode, Decode)]
+pub struct CompressedU64Diff {
+    bytes: Vec<u8>,
+}
+
+impl HDiffBuffer {
+    pub fn from_state<E: EthSpec>(mut beacon_state: BeaconState<E>) -> Self {
+        let _t = metrics::start_timer(&metrics::STORE_BEACON_HDIFF_BUFFER_FROM_STATE_TIME);
+        // Set state.balances to empty list, and then serialize state as ssz
+        let balances_list = std::mem::take(beacon_state.balances_mut());
+
+        let state = beacon_state.as_ssz_bytes();
+        let balances = balances_list.to_vec();
```

Comment:
We no longer need `balance_list`, I'm wondering if it's more efficient to convert it using `into_Iter()` rather than cloning? although I don't know how much differences this make given that overhead of copying a `u64` is minimal - although this list is really large.

Reply:
Oh great catch! I had assumed this was not cloning, but it is :facepalm:

I'll change it to `Vec::from(balances_list)` which will just move the vec out of the `VariableList`: https://docs.rs/ssz_types/0.8.0/src/ssz_types/variable_list.rs.html#133-137

Reply:
Actually nvm, I got confused :facepalm:

This is a `milhouse::List` (i.e. a tree) so we can't make it into a vec any more cheaply than this. And I think we need to keep the vec so that the diff computation remains fast

Reply:
Got it, thanks for looking into this